src/schedule/ll_schedule.c

```diff
@@ -94,13 +94,40 @@ static void schedule_ll_task_update_start(struct ll_schedule_data *sch,
 	task->start += next;
 }
 
+/* caller should hold the domain lock */
+static void schedule_ll_task_done(struct ll_schedule_data *sch,
+				  struct task *task)
+{
+	list_item_del(&task->list);
+
+	/* the task finished, decrease the count */
+	atomic_sub(&sch->domain->total_num_tasks, 1);
+
+	/* decrease task number of the core */
+	atomic_sub(&sch->num_tasks, 1);
+
+	/* the last task of the core, unregister the client/core */
+	if (!atomic_read(&sch->num_tasks) &&
```

Comment:
as the lock is hold by the caller, so even the atomicity is broken we are still fine, I change it for readability improvement, otherwise we have to use this as the atomic_xx() returns the original value:

```
x = atomic_sub(y);
if (!(x-1))
```

Reply:
> I don't think it improves readability at all. It adds confusion why a wrong use of an API is allowed. Either we're protected by some external locking, then we should remove `atomic_t` and just make it an `int` and make sure it's consistently protected, or we use `atomic_t` and we understand why we need it and we use it correctly.

We don't have to make sure the atomicity, but we still need the make them as volatile to make sure not reading them from cache as we need sync from L1 cache of different Cores, @lyakh what do you suggest here?

Reply:
IIRC the atomic types are used as there is or was a need to add tasks from within other tasks and the scheduler lock was already held at this point. Not sure if this is the case today.
@lyakh the Zephyr timer domain work you are doing **should** be a lot simpler wrt to locking/atomic since this is all done by Zephyr. We should see a significant reduction in complexity.